flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/GlobalStatisticsAggregatorTracker.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.shuffle;
+
+import java.io.Serializable;
+import org.apache.flink.annotation.Internal;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * GlobalStatisticsAggregator is used by {@link DataStatisticsCoordinator} to collect {@link
+ * DataStatistics} from {@link DataStatisticsOperator} subtasks for specific checkpoint. It stores
+ * the merged {@link DataStatistics} result and uses set to keep a record of all reported subtasks.
+ */
+@Internal
+class GlobalStatisticsAggregatorTracker<K> implements Serializable {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(GlobalStatisticsAggregatorTracker.class);
+  private static final double EXPECTED_DATA_STATISTICS_RECEIVED_PERCENTAGE = 90;
+  private final transient DataStatisticsFactory<K> statisticsFactory;
+  private final transient int parallelism;
+  private transient volatile GlobalStatisticsAggregator<K> inProgressAggregator;
+  private volatile GlobalStatisticsAggregator<K> lastCompletedAggregator;
+
+  GlobalStatisticsAggregatorTracker(DataStatisticsFactory<K> statisticsFactory, int parallelism) {
+    this.statisticsFactory = statisticsFactory;
+    this.parallelism = parallelism;
+  }
+
+  boolean receiveDataStatisticEventAndCheckCompletion(int subtask, DataStatisticsEvent<K> event) {
+    LOG.info("!!! Receive event {} from subtask {}", event, subtask);
```

Comment:
Nit: please fix the log message

Reply:
for debugging purpose, I will remove it.